### PR TITLE
Sanitize dot-graph output

### LINKED ISSUE #4328: Dot graph exporter `debug/dag/` does not escape label values
**What steps did you take and what happened:**
`CONTOUR_POD=$(kubectl -n envoyweb get pod -l app=contour -o jsonpath='{.items[0].metadata.name}')`  
`kubectl -n envoyweb port-forward $CONTOUR_POD 6060`  
`curl -s localhost:6060/debug/dag | dot`
> Error: <stdin>: syntax error in line 5 near ':'

I found there's special characters (`"`, `<`, `>`) in label values carried over from our listener config.

**What did you expect to happen:**

Output from the Graphql `dot` rendering tool should help visualize the contour DAG.

**Anything else you would like to add:**

**Environment:**

- Contour version: Fork of v0.12.1
- Kubernetes version: (use `kubectl version`): N/A
- Kubernetes installer & version: N/A
- Cloud provider or hardware configuration: N/A
- OS (e.g. from `/etc/os-release`): N/A
- `dot` version: dot - graphviz version 2.50.0 (20211204.2007)

### PR DESCRIPTION
**Pull request by kb000**
This change html-escapes text fields in the `/debug/dag` `dot` graph output.

The dot-graph output of contour's dag state can allow HTML-unsafe characters, which leads to failures rendering the graph in the `dot` utility.   (I found this issue in my enterprise's fork, where the issue is more severe, but I figured I'd upstream the fix, too.)

Added unit tests and refactored a bit to allow unit testing.

Fixes #4328

### CODE REVIEW

**Comment by kb000** [comment]
Addresses https://github.com/projectcontour/contour/issues/4328

**Comment by skriss** [comment]
Thanks for the PR @kb000! Will review soon.

File: internal/dag/builder.go [hunk 1]
```diff
@@ -35,6 +35,11 @@ func (pf ProcessorFunc) Run(dag *DAG, source *KubernetesCache) {
 	}
 }
 
+type BuilderType interface {
+	// Build builds and returns a new DAG
+	Build() *DAG
+}
+
```
Comment (skriss) [inline internal/dag/builder.go:42]: I might move this into `internal/debug` where it's required, and then can rename to `DAGBuilder` or similar.

File: internal/debug/dot.go [hunk 2]
```diff
@@ -24,21 +25,33 @@ import (
...
 
-	nodes := map[interface{}]bool{}
-	edges := map[pair]bool{}
+	printNodes(nodes, w)
+	printEdges(edges, w)
+
+	fmt.Fprintln(w, "}")
+}
+
+func collectDag(b dag.BuilderType) (nodes nodeCollection, edges edgeCollection) {
```
Comment (skriss) [inline internal/debug/dot.go:49]: we generally don't use named return parameters, let's drop the names here and declare them below.

**Review by skriss — CHANGES_REQUESTED** [review]
This looks great to me @kb000, just a couple small nits and I think we can get it in. Thanks again for the PR!

File: internal/dag/builder.go (hunk 1 above)
Comment (kb000) [reply internal/dag/builder.go:42]: Great suggestion. Duck interfacing blowing my mind here.

**Review by skriss — APPROVED** [review]

**Review by youngnick — APPROVED** [review]
LGTM, I think that the use of `mockery` is new, but maybe we should consider it for more places?

**Comment by sunjayBhatia** [comment]
> LGTM, I think that the use of `mockery` is new, but maybe we should consider it for more places?

Yeah we added it recently and this was nice to see additional usage from a contributor 🤙🏽

**Pull request merged**